realtek: add support for Netgear GS750E

DavideFioravanti opened the pull request:
This is an RTL8393-based switch with 48 RJ-45 and 2 SFP ports.

Hardware
--------
```
SoC:	Realtek RTL8393M
RAM:	128MB DDR3 (Nanya NT5CC64M16GP-DI)
FLASH:	8MB NOR (Macronix MX25L6433F)
ETH:	48x 10/100/1000 Mbps RJ-45 Ethernet
SFP:	2x SFP
BTN:
  - 1x	Reset button
LEDS:
  - 50x	Green-Amber leds: lan/sfp status
  - 1x	Green led: power (Always on)
UART:
  - 115200-8-N-1 (CN3, pin-out on PCB)
```

Everything works correctly except for the 2 SFP ports that are not
working unless you enable it every boot in U-Boot with the command:
`	rtk network on`

Installation
------------
You can install Openwrt using one of the following methods.

Warning: flashing OpenWrt will delete your current configuration.
Warning 2: if the -factory.bix file is not available anymore, you must
follow Method 2.

Method 1:
Check the firmware version currently running on your switch. If you are
running FW V1.0.1.10 or greater, you have to download the firmware
V1.0.1.8 from Netgear website and then flash this version. When the
switch restarts, it should be on version V1.0.1.8. Now you can get the
OpenWrt -factory.bix file and then flash it using the OEM web interface.

Method 2 (requires the UART connection):
Boot the -initramfs-kernel.bin image from U-Boot with these commands:
	rtk network on;
	tftpboot 0x8f000000 openwrt-realtek-rtl839x-netgear_gs750e-initramfs-kernel.bin;
	bootm;
And then flash the -sysupgrade.bin file from OpenWrt.

Revert to stock
---------------
Get the stock firmware from the Netgear website and flash it using the
OpenWrt web interface. Remember to not keep the current configuration
and check the "Force upgrade" checkbox

Once reverted to stock the firmware could complain in the UART console
about mtdblock3 and/or mtdblock4 not being mounted correctly but it
seems to work anyway without any problems. Sample error:
`	mount: Mounting /dev/mtdblock4 on /mntlog failed: Input/output error`
	
If you want to get rid of these error messages you can boot the
initramfs image from U-Boot with these commands:
```
	rtk network on
	tftpboot 0x8f000000 openwrt-realtek-rtl839x-netgear_gs750e-initramfs-kernel.bin;
	bootm;
```
	
And then erase the corresponding partitions using the command:
	For mtdblock3: 
		`mtd erase jffs2_cfg`
	For mtdblock4: 
		`mtd erase jffs2_log`

Now you can reboot the switch and the errors should be gone.

Note
----
To get the SFP ports fully working, all the right GPIOs must be found.
In the GPL sources I found these:
 - GPIO_14: SFP_TX_DIS1;
 - GPIO_19: SFP_TX_DIS0;

DavideFioravanti committed:
realtek: add support for Netgear GS750E

This is an RTL8393-based switch with 48 RJ-45 and 2 SFP ports.

Hardware
--------
SoC:	Realtek RTL8393M
RAM:	128MB DDR3 (Nanya NT5CC64M16GP-DI)
FLASH:	8MB NOR (Macronix MX25L6433F)
ETH:	48x 10/100/1000 Mbps RJ-45 Ethernet
SFP:	2x SFP
BTN:
  - 1x	Reset button
LEDS:
  - 50x	Green-Amber leds: lan/sfp status
  - 1x	Green led: power (Always on)
UART:
  - 115200-8-N-1 (CN3, pin-out on PCB)

Everything works correctly except for the 2 SFP ports that are not
working unless you enable it every boot in U-Boot with the command:
	rtk network on

Installation
------------
You can install Openwrt using one of the following methods.

Warning: flashing OpenWrt will delete your current configuration.
Warning 2: if the -factory.bix file is not available anymore, you must
follow Method 2.

Method 1:
Check the firmware version currently running on your switch. If you are
running FW V1.0.1.10 or greater, you have to download the firmware
V1.0.1.8 from Netgear website and then flash this version. When the
switch restarts, it should be on version V1.0.1.8. Now you can get the
OpenWrt -factory.bix file and then flash it using the OEM web interface.

Method 2 (requires the UART connection):
Boot the -initramfs-kernel.bin image from U-Boot with these commands:
	rtk network on;
	tftpboot 0x8f000000 openwrt-realtek-rtl839x-netgear_gs750e-initramfs-kernel.bin;
	bootm;
And then flash the -sysupgrade.bin file from OpenWrt.

Revert to stock
---------------
Get the stock firmware from the Netgear website and flash it using the
OpenWrt web interface. Remember to not keep the current configuration
and check the "Force upgrade" checkbox

Once reverted to stock the firmware could complain in the UART console
about mtdblock3 and/or mtdblock4 not being mounted correctly but it
seems to work anyway without any problems. Sample error:
	mount: Mounting /dev/mtdblock4 on /mntlog failed: Input/output error

If you want to get rid of these error messages you can boot the
-initramfs-kernel.bin image from U-Boot with these commands:
	rtk network on;
	tftpboot 0x8f000000 openwrt-realtek-rtl839x-netgear_gs750e-initramfs-kernel.bin;
	bootm;

And then erase the corresponding partitions using the command:
	For mtdblock3:
		mtd erase jffs2_cfg
	For mtdblock4:
		mtd erase jffs2_log

Now you can reboot the switch and the errors should be gone

Note
----
To get the SFP ports fully working, all the right GPIOs must be found.
In the GPL sources I found these:
 - GPIO_14: SFP_TX_DIS1;
 - GPIO_19: SFP_TX_DIS0;

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

DavideFioravanti committed:
realtek: rtl839x: enable driver for virtual mtd_concat devices in config

Enable the driver for the rtl839x target. It's required at least for
Netgear GS750E

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+};
+
+&spi0 {
+	status = "okay";
```
Small nit: add a newline between "status" and "flash@0".

DavideFioravanti replied:
I will fix it, thanks!

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+			fixed-link {
+				speed = <1000>;
+				full-duplex;
+				pause;
```
Forgive me my ignorance, but why did you add "pause" here? The `SWITCH_SFP_PORT()` macro doesn't have this. Would it be useful to add it there instead?

svanheule reviewed (commented):
You could probably run `show tech-support` on your device, running OEM FW, to get a listing of the GPIOs. See for example the [GS310TP board config](https://svanheule.net/switches/gs310tp#vendor_board_definition).

Speaking of the board, can you add a pin-out description for the UART header?

Are the invalid jffs2_ partitions fixed by OEM FW when you for example save settings? If settings cannot be persisted without a user performing the `mtd erase` operation, this should be noted in the commit message.

Typically these switches running SDK-based FW will strip any data following the uImage in a provided image. Since your install instructions tell us to use the -sysupgrade.bin file, I assume this is not the case for the GS750E?

If the SFP cages can be made to work by enabling the network from the bootloader, then u-boot must have the required code to do this. Since this is a single GPL binary, Netgear is required to provided to sources for this. Did you check their [GPL archive](https://www.downloads.netgear.com/files/GPL/GS750E_V1.0.0.19_GPL_SRC.zip)?

svanheule commented:
And before I forget: thank you for working on this device :)

DavideFioravanti commented:
Thank @svanheule for your review! :slightly_smiling_face:  
First of all, this is my first attempt to port OpenWrt to a switch (and rtlxxxx in general). I have experience in the router space but I don't know the peculiarities of this platform. So treat me like a noob :wink:.

> You could probably run `show tech-support` on your device, running OEM FW, to get a listing of the GPIOs. See for example the [GS310TP board config](https://svanheule.net/switches/gs310tp#vendor_board_definition).

I tried but unfortunately it's not present in the OEM FW:
```
Plus Switch GS750E# show tech-support
Unknown command
Plus Switch GS750E# show 
  ac               access control
  arp              Show the IP ARP translation table
  bonjour          Shows bonjour advertisement status
  cable-diag       Cable Diagnostics
  custom           Custom Module configuration
  debugging        debugging information
  dos              DoS information
  history          list the last several history commands
  info             Basic information
  interfaces       Interface status and configuration
  ip               IP information
  lacp             LACP Configuration
  lag              Link Aggregation Group Configuration
  loop-prevention  Loop-prevention configuration
  mac              MAC configuration
  mirror           Mirror configuration
  qos              QoS configuration
  running-config   Running configuration
  sccd             SCCD information
  startup-config   Startup configuration
  storm-control    Storm control configuration
  users            Display information about users
  version          System hardware and software status
  video-vlan       Video VLAN configuration
  vlan             VLAN configuration
  voice-vlan       Voice VLAN configuration
```
I have to enable something? I even tried the command `show debugging` but it doesn't output anything.
This is the configure section:
```
Plus Switch GS750E(config)# 
  ac               access control
  bonjour          Enables globally MDNS advertisement and responding to 
                   MDNS queries
  custom           Custom Module configuration
  dos              DoS information
  do               To run exec commands in current mode
  eee              EEE configuration
  end              End current mode and change to enable mode
  exit             Exit current mode and down to previous mode
  interface        Select an interface to configure
  ip               IP information
  lacp             LACP Configuration
  loop-prevention  Loop-prevention configuration
  mirror           Mirror configuration
  no               Negate command
  password         Change user password
  qos              QoS configuration
  sccd             SCCD configuration
  storm-control    Storm control configuration
  system           System information
  video-vlan       Video VLAN configuration
  vlan             VLAN configuration
  voice-vlan       Voice VLAN configuration
  web              Web UI
```
----------
> Speaking of the board, can you add a pin-out description for the UART header?

I didn't add it because it's labelled directly on the board:
![image](https://user-images.githubusercontent.com/2470733/211149756-1f7b0f71-a8c2-49ad-95fd-f7c7fa9dc302.png)
Do you think I should add it anyway?

----------
> Are the invalid jffs2_ partitions fixed by OEM FW when you for example save settings? If settings cannot be persisted without a user performing the `mtd erase` operation, this should be noted in the commit message.

In my tests the second jffs2_ partition is never recreated automatically while the first one was always reformatted except for one time (but I am not able to recreate that condition). However as I said in the commit message the switch seems to works without any problems even without manual reformat. To be safe I added anyway the commands to reformat the partitions. However I will add in the commit message a warning about the fact that the settings will be lost after the OpenWrt flash (in the router space it happens almost always :smile:).
Just for clarification, I had to use these two partitions because of the small free space available, otherwise I wouldn't have used them.

-------------
 
> Typically these switches running SDK-based FW will strip any data following the uImage in a provided image. Since your install instructions tell us to use the -sysupgrade.bin file, I assume this is not the case for the GS750E?

Oh, thank you I didn't know this quirk. When I tried the installation starting from OEM fw version 1.0.1.7 it was working correctly. However after your message I tested gain with the latest OEM fw version (1.0.1.10) and it didn't work. Openwrt booted but it was complaining about missing rootfs partition. So I think that Netgear added this check afterwards in the development. I will fix it. :wink: 

-----------

> If the SFP cages can be made to work by enabling the network from the bootloader, then u-boot must have the required code to do this. Since this is a single GPL binary, Netgear is required to provided to sources for this. Did you check their [GPL archive](https://www.downloads.netgear.com/files/GPL/GS750E_V1.0.0.19_GPL_SRC.zip)?

As I said in a previous comment, I looked for the pins in the GPL but I wasn't able to find them :disappointed:. 
I just easily found these pins in the uboot GPL:
In the file rtl8390_init.c:
// GPIO_10: System Reset PIN; GPIO_14: SFP_TX_DIS1;
// GPIO_19: SFP_TX_DIS0; GPIO_23: GPIO_RESET_PHY

From what I can see in other boards, these SFP should be connected to an RTL8231. Once defined in the dts:
```
	gpio1: rtl8231-gpio {
		compatible = "realtek,rtl8231-gpio";
		#gpio-cells = <2>;
		indirect-access-bus-id = <0>;
		gpio-controller;
	};
```
It gets detected:
```
[    0.407630] Probing RTL8231 GPIOs
[    0.411326] rtl8231_init called, MDIO bus ID: 0
```
But then I have to find the right i2c pins. I looked around in the GPL but I didn't find any clear indication. Do you have any advice? (I never worked with i2c nor sfp)

[OT]
I also found this pin in the file netgear_gs750e_board.c:
`.phy.rstGpio.rstIntGpio.pin = 23,`
It should be used to reset the PHY. I should set it in every internal PHY?
[/OT]

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+	virtual_flash {
+		compatible = "mtd-concat";
+
+		devices = <&fwconcat0 &fwconcat1 &fwconcat2>;
```
I found a lot of devices (if not all) using my devices definition: https://github.com/openwrt/openwrt/search?q=devices+%3D+%3C%26fwconcat0 Which is the right one?

svanheule replied:
It seems this should be a list of node references, not a node reference with two extra cells. See for example: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=25e4f5220efead592c83200241e098e757d37e1f

In the end, the resulting binary device tree (DTB) will contain exactly the same bytes for both notations. There also isn't any formal upstream binding you could check against, because AFAICT it was never accepted.

DavideFioravanti replied:
Ok then, I will fix it! :wink:

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+
+			partition@0 {
+				label = "firmware";
+				reg = <0x0 0x0>;
```
Other devices didn't specified the full size, so I thought it wasn't mandatory. I will add it, thank you!

svanheule replied:
Probably works without it because the driver doesn't use the info, but I think it's good to provide consistent info anyway.

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "loader";
```
I forgot it, thanks!

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+			};
+
+			partition@80000 {
+				label = "bdinfo";
```
I will fix it, thanks!

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+&ethernet0 {
+	mdio: mdio-bus {
+		compatible = "realtek,rtl838x-mdio";
+		regmap = <&ethernet0>;
```
I will fix it, thanks!

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+			};
+
+			partition@90000 {
+				label = "sysinfo";
```
I will rename the partition.

This partition contains only these variables:
```
$ strings mtd2.sysinfo.bin 
!boardid=1
resetdefault=0
sn=__REDACTED__
resetdflt=1
```

I can confirm that the switch gets automatically the right mac address (XX:XX:XX:XX:XX:fd) which is the same on the label:
```
root@OpenWrt:/# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1504 qdisc mq state UP group default qlen 1000
    link/ether XX:XX:XX:XX:XX:fd brd ff:ff:ff:ff:ff:ff
    inet6 fe80::2a80:88ff:fe66:d6fd/64 scope link 
       valid_lft forever preferred_lft forever
3: lan01@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether XX:XX:XX:XX:XX:fd brd ff:ff:ff:ff:ff:ff
4: lan02@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether XX:XX:XX:XX:XX:fd brd ff:ff:ff:ff:ff:ff
5: lan03@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 02:00:01:00:00:02 brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
6: lan04@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 02:00:01:00:00:03 brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
...
```

DavideFioravanti replied:
UPDATE:
After I renamed the partitions I get this output:
```
root@OpenWrt:/# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1504 qdisc mq state UP group default qlen 1000
    link/ether XX:XX:XX:XX:XX:fd brd ff:ff:ff:ff:ff:ff
3: lan1@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 2a:XX:XX:XX:XX:fe brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
4: lan2@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 2a:XX:XX:XX:XX:ff brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
5: lan3@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 2a:XX:XX:XX:XX:00 brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
6: lan4@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master switch state LOWERLAYERDOWN group default qlen 1000
    link/ether 2a:XX:XX:XX:XX:01 brd ff:ff:ff:ff:ff:ff permaddr XX:XX:XX:XX:XX:fd
```

DavideFioravanti commented on target/linux/realtek/image/rtl839x.mk:
```diff
@@ -9,6 +9,15 @@ define Device/d-link_dgs-1210-52
 
+define Device/netgear_gs750e
+  SOC := rtl8393
+  IMAGE_SIZE := 6528k
```
I can confirm that it doesn't support lzma:
```
Loading ...

   Version:   MIPS OpenWrt Linux-5.10.161
   Created:   2023-01-05  16:17:24 UTC
   Size:      5157221 Bytes = 4.9 MB
   Checksum ... OK
Unimplemented compression type 3
```

However I was thinking that we can define a factory image using the max size of 6528k and a sysupgrade size of 7552k (the sum of the 3 concatenated devices). So that when the factory image won't fit anymore, we can still continue to use the sysupgrade image while the new installation can still be performed using the "tftpboot initramfs -> flash from OpenWrt" method:
```
# When the factory image won't fit anymore, it can be removed.
# New installation will be performed booting from initramfs
# and then flashing from OpenWrt
define Device/netgear_gs750e
  SOC := rtl8393
  IMAGE_SIZE := 7552k
  FACTORY_SIZE := 6528k
  DEVICE_VENDOR := NETGEAR
  DEVICE_MODEL := GS750E
  UIMAGE_MAGIC := 0x174e4741
  IMAGES += factory.bin
  IMAGE/factory.bin := \
    append-kernel | \
    pad-to 64k | \
    append-rootfs | \
    pad-rootfs | \
    check-size $$$$(FACTORY_SIZE)
endef
TARGET_DEVICES += netgear_gs750e
```
What do you think about it? I will also update the commit message accordingly

svanheule replied:
For a factory image, you would need to create an initramfs with limited size. So if the initramfs image is also used for factory installs, then that could probably use a size check. (Meandering thought: maybe we should actually do that by default for all realtek devices...)

DavideFioravanti replied:
Unfortunately the initramfs image is even bigger than the sysupgrade image... Probably because the sysupgrade has the squashfs partition that has a better compression than then the normal gzip? In the meantime I added the factory.bix image and updated the installation section of the commit message.

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+		gpio-hog;
+		line-name = "System Reset PIN";
+		gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
+		output-low;
```
Yes, i will fix it. Thanks!

DavideFioravanti commented on target/linux/realtek/dts-5.10/rtl8393_netgear_gs750e.dts:
```diff
@@ -0,0 +1,281 @@
+		SWITCH_PORT(6, 07, qsgmii)
+		SWITCH_PORT(7, 08, qsgmii)
+
+		SWITCH_PORT(8, 09, qsgmii)
```
Sorry, I will fix it, thanks!

svanheule commented:
> > You could probably run `show tech-support` on your device, running OEM FW, to get a listing of the GPIOs. See for example the [GS310TP board config](https://svanheule.net/switches/gs310tp#vendor_board_definition).
> 
> I tried but unfortunately it's not present in the OEM FW:

If this device was based on an older SDK like some other switches, this information is probably just not present (or we haven't found it). 


> > Speaking of the board, can you add a pin-out description for the UART header?
> 
> I didn't add it because it's labelled directly on the board:

OK, then you don't need to add the pin-out to the commit message. Maybe "C3, pin-out on PCB".


> However I will add in the commit message a warning about the fact that the settings will be lost after the OpenWrt flash (in the router space it happens almost always).

Thanks, less surprises is usually better in this case. :)

> Just for clarification, I had to use these two partitions because of the small free space available, otherwise I wouldn't have used them.

Figured as much, which is why I didn't comment on it.

> I just easily found these pins in the uboot GPL: In the file rtl8390_init.c:
> 
> ```
> // GPIO_10: System Reset PIN; GPIO_14: SFP_TX_DIS1;
> // GPIO_19: SFP_TX_DIS0; GPIO_23: GPIO_RESET_PHY
> ```
> 
> From what I can see in other boards, these SFP should be connected to an RTL8231. Once defined in the dts:
> 
> ...
> 
> But then I have to find the right i2c pins. I looked around in the GPL but I didn't find any clear indication. Do you have any advice? (I never worked with i2c nor sfp)
> 
> [OT] I also found this pin in the file netgear_gs750e_board.c: `.phy.rstGpio.rstIntGpio.pin = 23,` It should be used to reset the PHY. I should set it in every internal PHY? [/OT]

RTL839x chips have plenty of integrated GPIO pins, so everything might be wired directly to the SoC. I've had a look at the GPL archive too, and it looks like the port LEDs are controlled in "bi-color scan mode", which means an RTL8231 is dedicated to the LEDs. If there is only one RTL8231 on your board, then I think it's unlikely to be used for GPIO control.

The phy reset GPIO should be added on the mdio node if it's shared. Do note that if you add it to the DTS, it might break ethernet functionality. I'm not sure how well the code handles bringing up the phy-s yet. If adding a `reset-gpios` line breaks your device, then it is the correct reset line however, so leave it commented out as a form of documentation. That way it could be enabled in the future if we ever figure it out.

GPIO 23 and 10 appear to be correct, so 14 and 19 are probably also correct. Usually they use "close by" GPIO lines for a single SFP cage.

If the SFP cages use SoC GPIO lines, then it might be quite hard to trace them out with a multimeter, and you may have to resort to some trial and error. You could e.g. toggle GPIO lines and check for level changes on pads or pull resistors near the SFP cages.

DavideFioravanti commented:
> RTL839x chips have plenty of integrated GPIO pins, so everything might be wired directly to the SoC. I've had a look at the GPL archive too, and it looks like the port LEDs are controlled in "bi-color scan mode", which means an RTL8231 is dedicated to the LEDs. If there is only one RTL8231 on your board, then I think it's unlikely to be used for GPIO control.

Actually there are 3 RTL8231 (I don't know if there are others below the heatsinks) and one of them is near the sfp ports:
![GS750E_INTERNAL (1)](https://user-images.githubusercontent.com/2470733/211224410-b22c5ffd-149f-48e4-918f-36c6087f0f92.jpg)
Closeup:
![GS750E_CLOSEUP](https://user-images.githubusercontent.com/2470733/211224544-44b2d37d-42a9-4a15-b034-2c14d8b84907.jpg)
As you can see some traces starts from the RTL8231 and ends to he resistors near the sfp. It's not a certainty, but I think that it's a strong evidence


> The phy reset GPIO should be added on the mdio node if it's shared. Do note that if you add it to the DTS, it might break ethernet functionality. I'm not sure how well the code handles bringing up the phy-s yet. If adding a `reset-gpios` line breaks your device, then it is the correct reset line however, so leave it commented out as a form of documentation. That way it could be enabled in the future if we ever figure it out.

I am testing it now. I will update this comment...

> GPIO 23 and 10 appear to be correct, so 14 and 19 are probably also correct. Usually they use "close by" GPIO lines for a single SFP cage.
> 
> If the SFP cages use SoC GPIO lines, then it might be quite hard to trace them out with a multimeter, and you may have to resort to some trial and error. You could e.g. toggle GPIO lines and check for level changes on pads or pull resistors near the SFP cages.

Unfortunately I will not be probably able to disassemble and test it with a multimeter in the following days... :disappointed: I'll let you know. However I think it can be also fixed later. Maybe we can continue the development on the forum? Or do you think that is better to pause this commit until it's fixed?

svanheule commented:
> Actually there are 3 RTL8231 (I don't know if there are others below the heatsinks) and one of them is near the sfp ports:

> As you can see some traces starts from the RTL8231 and ends to the resistors near the sfp. It's not a certainty, but I think that it's a strong evidence

Most of those traces terminating at the SFP cages seem to pass by U20, rather than originate from it. Following that bundle, they do actually appear to be SoC GPIO lines.

An RTL8231 can [support 24 ports](https://lore.kernel.org/lkml/22886e00f20e34338136e2e7afea90e6c87f9d68.1623532208.git.sander@svanheule.net/) in bi-color scan mode, so the board should have 3 to manage all 50 ports. If I'm not mistaken, the SoC will hardcode these at MDIO addresses 0, 1 and 2. The same MDIO bus is used to control the RTL8231-s from software. So the driver will probably see them if they are added to the device tree, but they would probably not function as expected.

If you can't figure out the SFP cages yet, it's OK to leave them unimplemented if you put it as a note in the commit message. We've done this before, where people may add the correct pin-outs later.

> I added it in the mdio-bus
> 
> ```
> mdio: mdio-bus {
> 		compatible = "realtek,rtl838x-mdio";
> 		#address-cells = <1>;
> 		#size-cells = <0>;
> 		reset-gpios = <&gpio0 23 GPIO_ACTIVE_LOW>;
> 
> 		/* External phy RTL8218B #1 */
> 		EXTERNAL_PHY(0)
> 		...
> ```
> 
> but it's still working. How I can test if this gpio is doing something?

That would make sense if the phy-s are properly managed by the driver. :sweat_smile: 
If you don't add them to the DT as a reset GPIO, but pull gpio23 to GND (either from userspace or as a gpio-hog), the phy-s should stop responding when in reset.

DavideFioravanti commented:
> Most of those traces terminating at the SFP cages seem to pass by U20, rather than originate from it. Following that bundle, they do actually appear to be SoC GPIO lines.
> 
> An RTL8231 can [support 24 ports](https://lore.kernel.org/lkml/22886e00f20e34338136e2e7afea90e6c87f9d68.1623532208.git.sander@svanheule.net/) in bi-color scan mode, so the board should have 3 to manage all 50 ports. If I'm not mistaken, the SoC will hardcode these at MDIO addresses 0, 1 and 2. The same MDIO bus is used to control the RTL8231-s from software. So the driver will probably see them if they are added to the device tree, but they would probably not function as expected.
> 
> If you can't figure out the SFP cages yet, it's OK to leave them unimplemented if you put it as a note in the commit message. We've done this before, where people may add the correct pin-outs later.

Ok, then I made a wrong assumption. Thank you very much for your explanation! :slightly_smiling_face: 
It was already specified that the SFP was not working unless enabled from U-Boot. Now I added a note at the end of the commit message about the GPIOs missing and the two GPIOs already found for future development.

> That would make sense if the phy-s are properly managed by the driver. sweat_smile If you don't add them to the DT as a reset GPIO, but pull gpio23 to GND (either from userspace or as a gpio-hog), the phy-s should stop responding when in reset.

That GPIO is clearly doing something. I tried setting the GPIO 23 high and it broke the ethernet:
```
root@OpenWrt:/# sh /tmp/gpio.sh 23 out 1
root@OpenWrt:/# [   49.245886] rtl83xx-switch switch@1b000000 lan13: Link is Down
[   49.252545] rtl83xx-switch switch@1b000000 lan13: Link is Up - 1Gbps/Full - flow control rx/tx
[   49.262899] switch: port 13(lan13) entered disabled state
[   49.269630] switch: port 13(lan13) entered blocking state
[   49.275769] switch: port 13(lan13) entered forwarding state
[   49.596120] rtl83xx-switch switch@1b000000 lan13: Link is Down
[   49.602913] switch: port 13(lan13) entered disabled state
```
After a reboot I tried setting the GPIO 23 low and it crashed the driver and then broke the ethernet:
```
root@OpenWrt:/# sh /tmp/gpio.sh 23 out 0
root@OpenWrt:/# [   58.723919] Realtek RTL8218B (external) mdio-bus:2a: Master/Slave resolution failed, maybe conflicting manual settings?
[   58.732399] rtl83xx-switch switch@1b000000 lan13: Link is Down
[   58.736060] ------------[ cut here ]------------
[   58.743662] rtl83xx-switch switch@1b000000 lan13: Link is Up - 1Gbps/Full - flow control rx/tx
[   58.747761] WARNING: CPU: 0 PID: 4146 at drivers/net/phy/phy.c:963 phy_state_machine+0xa0/0x2dc
[   58.767260] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[   58.838954] CPU: 0 PID: 4146 Comm: kworker/0:5 Not tainted 5.10.162 #0
[   58.846528] Workqueue: events_power_efficient phy_state_machine
[   58.853138] Stack : 807d0e1c fffffffe 82b57d54 8017fd3c 807a9d88 80860000 82b52c80 807b0000
[   58.862490]         00000009 00000000 00000000 8017fedc 80862c07 00000001 82b57d10 f4054c4c
[   58.871836]         00000000 00000000 807a9d88 82b57b90 ffffefff 00000000 00000000 ffffffea
[   58.881183]         00000213 82b57b9c 00000213 80868090 00000000 00000001 00000000 807b0000
[   58.890528]         00000009 00000000 00000000 fffffffe 00000018 80463120 00000000 81620000
[   58.899869]         ...
[   58.902594] Call Trace:
[   58.905339] [<801070b8>] show_stack+0x30/0x100
[   58.910304] [<80405b64>] dump_stack+0x9c/0xcc
[   58.915177] [<8012c838>] __warn+0xc0/0x12c
[   58.919741] [<8012c900>] warn_slowpath_fmt+0x5c/0xac
[   58.925287] [<804bac50>] phy_state_machine+0xa0/0x2dc
[   58.930938] [<801472e4>] process_one_work+0x208/0x46c
[   58.936576] [<80147b74>] worker_thread+0x180/0x5dc
[   58.941929] [<8014d9fc>] kthread+0x13c/0x144
[   58.946695] [<80101c58>] ret_from_kernel_thread+0x14/0x1c
[   58.952718] 
[   58.954810] ---[ end trace 383d68c4aec9cf46 ]---
[   58.960491] switch: port 13(lan13) entered disabled state
[   58.960952] Realtek RTL8218B (external) mdio-bus:1a: Master/Slave resolution failed, maybe conflicting manual settings?
[   58.968180] switch: port 13(lan13) entered blocking state
[   58.978733] ------------[ cut here ]------------
[   58.984789] switch: port 13(lan13) entered forwarding state
[   58.989940] WARNING: CPU: 0 PID: 4233 at drivers/net/phy/phy.c:963 phy_state_machine+0xa0/0x2dc
[   59.005988] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[   59.077704] CPU: 0 PID: 4233 Comm: kworker/0:6 Tainted: G        W         5.10.162 #0
[   59.086842] Workqueue: events_power_efficient phy_state_machine
[   59.093458] Stack : 807d0e1c fffffffe 82b4fd54 8017fd3c 807a9d88 80860000 82b52600 807b0000
[   59.102810]         00000009 00000000 00000000 8017fedc 80862c07 00000001 82b4fd10 f4054c4c
[   59.112155]         00000000 00000000 807a9d88 82b4fb90 ffffefff 00000000 00000000 ffffffea
[   59.121502]         0000022e 82b4fb9c 0000022e 80868090 00000000 00000001 00000000 807b0000
[   59.130847]         00000009 00000000 00000000 fffffffe 00000018 80463120 00000000 81620000
[   59.140193]         ...
[   59.142917] Call Trace:
[   59.145658] [<801070b8>] show_stack+0x30/0x100
[   59.150622] [<80405b64>] dump_stack+0x9c/0xcc
[   59.155496] [<8012c838>] __warn+0xc0/0x12c
[   59.160070] [<8012c900>] warn_slowpath_fmt+0x5c/0xac
[   59.165604] [<804bac50>] phy_state_machine+0xa0/0x2dc
[   59.171257] [<801472e4>] process_one_work+0x208/0x46c
[   59.176895] [<80147b74>] worker_thread+0x180/0x5dc
[   59.182248] [<8014d9fc>] kthread+0x13c/0x144
[   59.187014] [<80101c58>] ret_from_kernel_thread+0x14/0x1c
[   59.193037] 
[   59.195139] ---[ end trace 383d68c4aec9cf47 ]---
[   59.201403] Realtek RTL8218B (external) mdio-bus:07: Master/Slave resolution failed, maybe conflicting manual settings?
[   59.213905] ------------[ cut here ]------------
[   59.219099] WARNING: CPU: 0 PID: 4080 at drivers/net/phy/phy.c:963 phy_state_machine+0xa0/0x2dc
[   59.229226] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[   59.301322] CPU: 0 PID: 4080 Comm: kworker/0:4 Tainted: G        W         5.10.162 #0
[   59.310473] Workqueue: events_power_efficient phy_state_machine
[   59.317078] Stack : 807d0e1c fffffffe 82babd54 8017fd3c 807a9d88 80860000 8219fb00 807b0000
[   59.326418]         00000009 00000000 00000000 8017fedc 80862c07 00000001 82babd10 f4054c4c
[   59.335763]         00000000 00000000 807a9d88 82babb90 ffffefff 00000000 00000000 ffffffea
[   59.345110]         00000246 82babb9c 00000246 80868090 00000000 00000001 00000000 807b0000
[   59.354455]         00000009 00000000 00000000 fffffffe 00000018 80463120 00000000 81620000
[   59.363800]         ...
[   59.366526] Call Trace:
[   59.369266] [<801070b8>] show_stack+0x30/0x100
[   59.374231] [<80405b64>] dump_stack+0x9c/0xcc
[   59.379104] [<8012c838>] __warn+0xc0/0x12c
[   59.383678] [<8012c900>] warn_slowpath_fmt+0x5c/0xac
[   59.389212] [<804bac50>] phy_state_machine+0xa0/0x2dc
[   59.394865] [<801472e4>] process_one_work+0x208/0x46c
[   59.400512] [<80147b74>] worker_thread+0x180/0x5dc
[   59.405856] [<8014d9fc>] kthread+0x13c/0x144
[   59.410623] [<80101c58>] ret_from_kernel_thread+0x14/0x1c
[   59.416639] 
[   59.418744] ---[ end trace 383d68c4aec9cf48 ]---
[   59.424559] Realtek RTL8218B (external) mdio-bus:08: Master/Slave resolution failed, maybe conflicting manual settings?
[   59.437059] ------------[ cut here ]------------
[   59.442369] WARNING: CPU: 0 PID: 4021 at drivers/net/phy/phy.c:963 phy_state_machine+0xa0/0x2dc
[   59.452495] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[   59.524618] CPU: 0 PID: 4021 Comm: kworker/0:3 Tainted: G        W         5.10.162 #0
[   59.533781] Workqueue: events_power_efficient phy_state_machine
[   59.540392] Stack : 807d0e1c fffffffe 822a5d54 8017fd3c 807a9d88 80860000 82b20b80 807b0000
[   59.549740]         00000009 00000000 00000000 8017fedc 80862c07 00000001 822a5d10 f4054c4c
[   59.559081]         00000000 00000000 807a9d88 822a5b90 ffffefff 00000000 00000000 ffffffea
[   59.568428]         0000025e 822a5b9c 0000025e 80868090 00000000 00000001 00000000 807b0000
[   59.577773]         00000009 00000000 00000000 fffffffe 00000018 80463120 00000000 81620000
[   59.587118]         ...
[   59.589843] Call Trace:
[   59.592588] [<801070b8>] show_stack+0x30/0x100
[   59.597550] [<80405b64>] dump_stack+0x9c/0xcc
[   59.602430] [<8012c838>] __warn+0xc0/0x12c
[   59.606995] [<8012c900>] warn_slowpath_fmt+0x5c/0xac
[   59.612539] [<804bac50>] phy_state_machine+0xa0/0x2dc
[   59.618181] [<801472e4>] process_one_work+0x208/0x46c
[   59.623821] [<80147b74>] worker_thread+0x180/0x5dc
[   59.629165] [<8014d9fc>] kthread+0x13c/0x144
[   59.633932] [<80101c58>] ret_from_kernel_thread+0x14/0x1c
[   59.639948] 
[   59.642061] ---[ end trace 383d68c4aec9cf49 ]---
[   59.684128] rtl83xx-switch switch@1b000000 lan13: Link is Down
[   59.690914] switch: port 13(lan13) entered disabled state
```
So I added a note in the dts: https://github.com/openwrt/openwrt/pull/11703/commits/2fe02c441b5f218689b391a789bff837ff9ec913#diff-13b9a91ab6bcfe637dbc3c7d628bafe2db0cdbd6feeb5cb7a5898055afa4578bR111

------

I updated and rebased on top of master

svanheule commented:
If your gpio.sh script is the one from the OpenWrt wiki, then `gpio.sh 23 high` will first set the GPIO to be an output, and then in a second step to be high. After the first step it will be momentarily set low however, which will trigger a reset on all you phy-s.

In your second test, setting the GPIO to a permanent low value keeps them in that reset state. Which apparently causes the driver to freak out a bit.

I would suggest you add the GPIO as a `reset-gpios` property, if that doesn't cause any issues.

DavideFioravanti commented:
> If your gpio.sh script is the one from the OpenWrt wiki, then `gpio.sh 23 high` will first set the GPIO to be an output, and then in a second step to be high. After the first step it will be momentarily set low however, which will trigger a reset on all you phy-s.
> 
> In your second test, setting the GPIO to a permanent low value keeps them in that reset state. Which apparently causes the driver to freak out a bit.

This is the script I was using: https://openwrt.org/docs/techref/hardware/port.gpio#utilities. I didn't know that setting the pin as output would momentarily set it in low state, this explains everything. Thank you again for  another explanation! :slightly_smiling_face: :+1: 

> I would suggest you add the GPIO as a `reset-gpios` property, if that doesn't cause any issues.

I added it [here](https://github.com/openwrt/openwrt/pull/11703/commits/b19a6c119a1b842d01fbfc2eca5a1615b00799c2#diff-13b9a91ab6bcfe637dbc3c7d628bafe2db0cdbd6feeb5cb7a5898055afa4578bR111), rebased on top of master and it's still working correctly. :slightly_smiling_face:

svanheule commented:
Thanks for the update, I've merged the changes with one minor tweak (dropped a newline in the dts files).

The `FACTORY_SIZE` variable used in the Makefile also hadn't been declared as a device variable. That means it keeps it's values across device recipes, but that probably wasn't an issue here (yet). Since it was already in use on the ath79 target too, I added a small patch in front of yours to fix that globally.
